main.nf

```diff
@@ -464,31 +464,55 @@ workflow {
         }
 
         GATHERBQSRREPORTS(recaltable)
-        // if ('baserecalibrator' in skip_qc) baseRecalibratorReport.close()
+        table = GATHERBQSRREPORTS.out.table
+    } else {
+        table = BASERECALIBRATOR.out.report
     }
 
     // STEP 4: RECALIBRATING
-    bam_applybqsr = MARKDUPLICATES.out.bam.join(GATHERBQSRREPORTS.out.table) //by:[0]
-    bam_applybqsr = bam_applybqsr.combine(BUILD_INDICES.out.intervals)
+    applybqsr_bam = markduplicates_bam.join(table)
+
+    applybqsr_bam = applybqsr_bam.combine(intervals)
         // if (step == 'recalibrate') bamApplyBQSR = input_sample
-    APPLYBQSR(bam_applybqsr, dict, fasta, fai)
+    APPLYBQSR(applybqsr_bam, dict, fasta, fai)
 
     // STEP 4.5: MERGING AND INDEXING THE RECALIBRATED BAM FILES
-    if (!(params.no_intervals)){
-        MERGE_BAM_RECAL(APPLYBQSR.out)
-        SAMTOOLS_INDEX_RECAL(MERGE_BAM_RECAL.out)
-    }else{
-        SAMTOOLS_INDEX_RECAL(APPLYBQSR.out)
+    if (!params.no_intervals) {
+        APPLYBQSR.out.map{ meta, bam -> //, bai ->
+            patient = meta.patient
+            sample  = meta.sample
+            gender  = meta.gender
+            status  = meta.status
+            [patient, sample, gender, status, bam] //, bai]
+        }.groupTuple(by: [0,1]).set{ bam_recal_to_merge }
+
+        bam_recal_to_merge = bam_recal_to_merge.map {
+            patient, sample, gender, status, bam -> //, bai ->
+
+            def meta = [:]
+            meta.patient = patient
+            meta.sample = sample
+            meta.gender = gender[0]
+            meta.status = status[0]
+            meta.id = sample
+
+            [meta, bam]
+        }
+
+        MERGE_BAM_RECAL(bam_recal_to_merge)
+        recal = MERGE_BAM_RECAL.out
+    } else {
+        recal = APPLYBQSR.out
     }
 
     // STEP 5: QC
-    if(!('samtools' in skip_qc))
-        SAMTOOLS_STATS(MERGE_BAM_RECAL.out)
-    //TODO This should work but somehow BAMQC is not called
-    bamqc = BWAMEM2_MEM.out//.mix(MERGE_BAM_RECAL.out)
-    //if(!('bamqc' in skipQC))
-        BAMQC(bamqc)//, target_bed)
+    if (!('samtools' in skip_qc)) {
+        SAMTOOLS_STATS(BWAMEM2_MEM.out.mix(recal))
+    }
 
+    if (!('bamqc' in skip_qc)) {
+        BAMQC(BWAMEM2_MEM.out.mix(recal), target_bed)
+    }
```

Comment:
Maybe for consistency we could always set the output to a variable, so above somewhere bwamem2_out = BWAMEM2_MEM.out ..... Maybe a more descriptive variable name.

Reply:
That's a good idea too.
Maybe something like `bwamem2_bam` or `mapped_bam`